### PR TITLE
rest: add verbose and mempool_sequence query params for mempool/contents

### PR DESCRIPTION
**Pull request by andrewtoth**
The verbose mempool json response can get very large. This adds an option to return the non-verbose response of just the txids. It is identical to the rpc response so the diff here is minimal. This also adds the mempool_sequence parameter for rpc consistency. Verbose defaults to true to remain backwards compatible.

It uses query parameters to be compatible with the efforts in https://github.com/bitcoin/bitcoin/issues/25752.

### CODE REVIEW

File: src/rest.cpp
Comment (brunoerg) [inline src/rest.cpp:622]: This error and the previous ones could be covered in the functional test

**Review by brunoerg — COMMENTED** [review]
Concept ACK

You should update the doc (`REST-interface.md`) as well

File: src/rest.cpp
Comment (brunoerg) [inline src/rest.cpp:611]: maybe use a `const` here?
```suggestion
            const std::string raw_verbose = req->GetQueryParameter("verbose").value_or("true");
```

**Comment by andrewtoth** [comment]
> You should update the doc (REST-interface.md) as well

@brunoerg would it make sense to update that file after there has been a release with this patch? I think that might confuse users who see those options in the doc but are unable to use them in any release.

File: src/rest.cpp
Comment (stickies-v) [inline src/rest.cpp:618]: I think the most elegant solution here would be to introduce a `HTTPRequest::GetQueryParameter<bool>` specialization that converts to and returns a bool, or throws an err if the string can't be parsed. However, that would require some more fundamental changes to how we deal with errors (which unfortunately we return instead of throw) here. Flagging it here for follow-up, feel free to grab it if you're keen.

File: src/rest.cpp
Comment (stickies-v) [inline src/rest.cpp:619]: {} list initialization is preferred (+ in other places):
```suggestion
            const bool verbose{raw_verbose == "true"};
```

**Review by stickies-v — COMMENTED** [review]
Concept ACK - bringing it further in line with the `getrawmempool` RPC seems reasonable.

> would it make sense to update that file after there has been a release with this patch?

To minimize maintenance I'd suggest labeling this pull with the 25.0 milestone and adding a small disclaimer to the `rest-interface.md` documentation that these 2 parameters are available only from 25.0, similar to https://github.com/bitcoin/bitcoin/blob/3baa0f5a603cedd66442492346583f2b3d776db7/doc/REST-interface.md?plain=1#L56
That way when this PR doesn't make it into 25.0 it'll be easy to track and update the docs.

File: src/rest.cpp
Comment (brunoerg) [reply src/rest.cpp:618]: Strong concept ACK on introducing a `HTTPRequest::GetQueryParameter<bool>`, it would be more elegant than using string ("true"/"false"), but I know it would require a lot of more work and maybe could be done in a follow-up.

Comment (andrewtoth) [reply src/rest.cpp]: I can definitely do that in a follow-up PR after this is merged.

**Comment by andrewtoth** [comment]
> I'd suggest labeling this pull with the 25.0 milestone

@stickies-v I don't think I have permissions to add labels. Otherwise I have addressed all feedback from yourself and @brunoerg .

File: doc/REST-interface.md
Comment (stickies-v) [inline doc/REST-interface.md:131]:
```suggestion
`verbose=true` and `mempool_sequence=false`.
```

File: doc/REST-interface.md
Comment (stickies-v) [inline doc/REST-interface.md:126]: Actually, since `false` is default, perhaps better to set that as the first option?
```suggestion
`GET /rest/mempool/contents.json?verbose=<true|false>&mempool_sequence=<false|true>`
```

**Review by stickies-v — APPROVED** [review]
ACK c35c7d091 (for 25.0)

> I don't think I have permissions to add labels.

The maintainers should be able to add the 25.0 label when they see this, otherwise you can always send a message on #bitcoin-core-dev IRC. It's not super urgent though, just would be a bit easier to track I think.

File: doc/REST-interface.md
Comment (brunoerg) [inline doc/REST-interface.md:131]:
```suggestion
`verbose=true` and `mempool_sequence=false`.
```

Comment (andrewtoth) [reply doc/REST-interface.md]: Ahh sorry should have caught that.

**Review by stickies-v — APPROVED** [review]
re-ACK [1ff5d61](https://github.com/bitcoin/bitcoin/pull/26207/commits/1ff5d61dfdaf8987e5619162662e4c760af76a43)

I verified that the only changes since c35c7d0 are suggested documentation updates to `REST-interface.md`.

**Review by pablomartin4btc — APPROVED** [review]
tested ACK 1ff5d61dfdaf8987e5619162662e4c760af76a43.

I've performed manual tests getting the node started with `-rest`, comparing results between the rest api call using curl with the different combinations of params and values vs. calling the `getrawtransacion` rpc command (using `bitcoin-cli`), which is essentially what the python functional tests do.

It works as expected and it's consistent with rpc.

**Comment by achow101** [comment]
ACK 1ff5d61dfdaf8987e5619162662e4c760af76a43

**Comment by jamesmichaeltate1** [comment]
(Bitcoin.org)

**Pull request merged**
